### ISSUE TITLE
onPickupLeave+setDimension

### ISSUE THREAD

**Issue by Allerek** [multitheftauto/mtasa-blue#846]
onPickupLeave do not detect the dimension change,idk how about interior.

**Comment by botder** [comment]
Can you use the **Bug report** issue template next time? Your description is lacking information.
How to reproduce your bug? What is the expected behavior? What is your server/client version?

**Comment by Allerek** [comment]
> 
> 
> Can you use the **Bug report** issue template next time? Your description is lacking information.
> How to reproduce your bug? What is the expected behavior? What is your server/client version?
1. Create pickup
2. Stand in it, then teleport to other dimension.
I have the script that makes GUI appear when you hit Pickup the triggerClientEvent with  function when you leave the pickup thats hiding the GUI, and it doesnt hide if you change dimension when standing in pickup.
Client/Server version-Latest Public

**Comment by Allerek** [comment]
https://youtu.be/4P1k4VyohiM

**Comment by botder** [comment]
# Client
https://github.com/multitheftauto/mtasa-blue/blob/d47332e88ef0e7285a9f7f257710fe864424ff74/Client/mods/deathmatch/logic/CClientPickup.cpp#L196-L214

# Server
https://github.com/multitheftauto/mtasa-blue/blob/d47332e88ef0e7285a9f7f257710fe864424ff74/Server/mods/deathmatch/logic/CPickup.cpp#L514-L541

# Proposed solution
Change `CPickup::Callback_OnLeave` (**server**) to match the behavior of `CClientPickup::Callback_OnLeave` (**client**):
- Add _matchingDimensions_ parameter
- Trigger for leaving players in different interior and dimension

We could also do the same for `CPickup::Callback_OnCollision` (client triggers onClientPickupHit even if interior and dimension differs, but server doesn't), but it might break existing scripts.

**Comment by qaisjp** [comment]
Actually I think it's good that onPickupLeave server-side currently only triggers for matching interiors and dimensions. 

I am not sure what to do if the user leaves a pickup by teleporting to another dimension. It should probably trigger the event even though the dimension/interior is different, as the player has technically left the pickup.

I have a hunch that the clientside argument was only added in the name of backwards compatibility

Thoughts @botder?

**Comment by botder** [comment]
I really dislike keeping backwards compatibility in such cases, because the next scripter might not expect this odd behavior - he probably expects that leaving the pickup's dimension/interior should trigger the event.

**Comment by qaisjp** [comment]
> - Add matchingDimensions parameter [the first half]
> - Trigger for leaving players in different interior and dimension [the second half]

> I really dislike keeping backwards compatibility in such cases, because the next scripter might not expect this odd behavior - he probably expects that leaving the pickup's dimension/interior should trigger the event.

I agree we should add a fix for the second half. My backwards compatibility comment is only relating to the first half.

> I have a hunch that the clientside argument was only added in the name of backwards compatibility

This meant that the client-side event probably originally used to always trigger (regardless of dimension and interior). So they added `matchingDimension` for people to check instead of breaking BC a little.

My argument is that _it is good that onPickupLeave does not trigger when you pickup is in a different dimension_. i.e., we should not remove the dimension/interior checks on the server side.

Also, on the client, `matchingDimensions` should probably be `true` if the player leaves the pickup's dimension, since the player USED TO BE in the same dimension.

Does that make sense?

**Comment by Allerek** [comment]
Refresh, is it fixed?
